File: matlab/thomas_algorithm.m
A = [
    [390, 201, 0, 0, 0, 0, 0, 0, 0, 0],
    [21, -135, -53, 0, 0, 0, 0, 0, 0, 0],
    [0, -271, 868, -203, 0, 0, 0, 0, 0, 0],
    [0, 0, -7, 596, 160, 0, 0, 0, 0, 0],
    [0, 0, 0, -47, -208, -96, 0, 0, 0, 0],
    [0, 0, 0, 0, -24, 347, -17, 0, 0, 0],
    [0, 0, 0, 0, 0, -96, 220, -40, 0, 0],
    [0, 0, 0, 0, 0, 0, 107, 546, -204, 0],
    [0, 0, 0, 0, 0, 0, 0, -301, -740, -364],
    [0, 0, 0, 0, 0, 0, 0, 0, 48, -154],
];

y = [
    [446],
    [62],
    [-783],
    [264],
    [-747],
    [-732],
    [-803],
    [-716],
    [-664],
    [-608],
];

function [roots, det] = thomas(A, y)
    n = size(A, 1);

    coeff = zeros(n, 2);
    det = 1;
    alpha = 0;
    beta = 0;

    for i = 1:n
    	if(i - 1 > 0)
        	b = 1. * A(i, i - 1);
        else
        	b = 0.;
        endif

        c = 1. * A(i, i);
        
        if(i != n)
        	d = 1. * A(i, i + 1);
        else
        	d = 0.;
        endif

        det *= (c + b * alpha);

        if c + b * alpha == 0
            disp('Correctness failed');
        end;
        tmp = -d / (c + b * alpha);
        if abs(tmp) > 1
            disp('UNSTABLE');
        end;
        beta = (y(i, 1) - b * beta) / (c + b * alpha);
        alpha = tmp;

        coeff(i, :) = [alpha, beta];
    end;

    coeff
    roots = [beta];

    for i = n - 1:-1:1
        alpha = coeff(i, 1);
    	beta = coeff(i, 2);
        roots = [alpha * roots(1) + beta, roots];
    end;
end;

tic
[roots, det] = thomas(A, y)
toc

File: python/thomas_algorithm.py
import time

from matrix_tool import print_vector, print_matrix
from check import check_roots

EPS = 5e-10


def solve(A, y):

  n = len(A)

  coeff = []
  det = 1
  alpha = 0
  beta = 0

  for i in xrange(n):
    b = (1. * A[i][i - 1]) if i - 1 >= 0 else 0.
    c = 1. * A[i][i]
    d = (1. * A[i][i + 1]) if i + 1 != n else 0.
    # Correctness check
    if abs(c + b * alpha) < EPS:
      print("Correctness failed")
      raise ValueError()
    det *= (c + b * alpha)
    alpha, beta = -d / (c + b * alpha), (y[i][0] - b * beta) / (c + b * alpha)
    # Stability check
    if abs(alpha) - 1. > EPS:
      print("UNSTABLE")
    coeff.append((alpha, beta))

  print("  alpha\t  beta")
  print_matrix([[a, b] for a, b in coeff])
  roots = [beta]

  for i in xrange(n - 2, -1, -1):
    alpha, beta = coeff[i]
    roots.insert(0, alpha * roots[0] + beta)

  return roots, det


def main():
  from constants import A3_21 as A, b3_21 as b
  start_time = time.time()
  x, det = solve(A, b)
  solve_time = time.time()
  print("Roots:")
  print_vector(x)
  check_roots(A, x, b)
  print
  print("Determinant: {}".format(det))

  print("Time elapsed: ")
  print("\tSolution:\t{} ms".format((solve_time - start_time) * 1000))


if __name__ == '__main__':
  main()